proxy/src/main/java/com/wavefront/agent/listeners/ZipkinPortUnificationHandler.java

```diff
@@ -0,0 +1,260 @@
+  private final Counter discardedTraces;
+  private final Counter discardedBatches;
+  private final Counter processedBatches;
+  private final Counter failedBatches;
+
+  private final static Set<String> ZIPKIN_VALID_PATHS = ImmutableSet.of("/api/v1/spans/", "/api/v2/spans/");
+  private final static String ZIPKIN_VALID_HTTP_METHOD = "POST";
+  private final static String APPLICATION_KEY = "application";
+  private final static String SERVICE_KEY = "service";
+  private final static String CLUSTER_KEY = "cluster";
+  private final static String SHARD_KEY = "shard";
+  private final static String SOURCE_KEY = "source";
+  private final static String DEFAULT_APPLICATION = "Zipkin";
+
+  public ZipkinPortUnificationHandler(String handle,
+                                      ReportableEntityHandlerFactory handlerFactory,
+                                      AtomicBoolean traceDisabled) {
+    this(handle, handlerFactory.getHandler(HandlerKey.of(ReportableEntityType.TRACE, handle)), traceDisabled);
+  }
+
+  public ZipkinPortUnificationHandler(final String handle,
+                                      ReportableEntityHandler<Span> handler,
+                                      AtomicBoolean traceDisabled) {
+    super(TokenAuthenticatorBuilder.create().setTokenValidationMethod(TokenValidationMethod.NONE).build(), handle,
+            false, true);
+    this.handle = handle;
+    this.handler = handler;
+    this.traceDisabled = traceDisabled;
+    this.discardedTraces = Metrics.newCounter(new MetricName("spans." + handle, "", "discarded"));
+    this.discardedBatches = Metrics.newCounter(new MetricName("spans." + handle + ".batches", "", "discarded"));
+    this.processedBatches = Metrics.newCounter(new MetricName("spans." + handle + ".batches", "", "processed"));
+    this.failedBatches = Metrics.newCounter(new MetricName("spans." + handle + ".batches", "", "failed"));
+  }
+
+  @Override
+  protected void handleHttpMessage(final ChannelHandlerContext ctx,
+                                   final FullHttpRequest incomingRequest) {
+    URI uri = parseUri(ctx, incomingRequest);
+    if (uri == null) return;
+
+    String path = uri.getPath().endsWith("/") ? uri.getPath() : uri.getPath() + "/";
+
+    // Validate Uri Path and HTTP method of incoming Zipkin spans.
+    if (!ZIPKIN_VALID_PATHS.contains(path)) {
+      writeHttpResponse(ctx, HttpResponseStatus.BAD_REQUEST, "Unsupported URL path.", incomingRequest);
+      logWarning("WF-400: Requested URI path '" + path + "' is not supported.", null, ctx);
+      return;
+    }
+    if (!incomingRequest.method().toString().equalsIgnoreCase(ZIPKIN_VALID_HTTP_METHOD)) {
+      writeHttpResponse(ctx, HttpResponseStatus.BAD_REQUEST, "Unsupported Http method.", incomingRequest);
+      logWarning("WF-400: Requested http method '" + incomingRequest.method().toString() + "' is not supported.", null, ctx);
+      return;
+    }
+
+    HttpResponseStatus status;
```

Comment:
I suggest you do this check - if (traceDisabled.get()) {
here instead of wasting CPU.
the traceDisabled flag is changed dynamically using feedback from wavefront server checkin method.

Reply:
The reason we do this further down is to be able to report "discarded no. of spans" metric values.

Reply:
ok, Removed tracking metrics for no. of discarded spans.